### PR TITLE
Allow a specific microphone to be used by the ffmpeg audio pipeline utility functions. Default to using the currently active microphone on Mac

### PR DESCRIPTION
**Pull request by jrhe**
Adds a `ffmpeg_input_device` parameter to `ffmpeg_microphone` and `ffmpeg_microphone_live` to specify the microphone to be used as input. 
Changes the default input device to be the currently active device on Mac/Darwin, rather than device 0.

Addresses issues #31820, #31074 , #28154, #31393.

I unfortunately can't get the tests pass on my machine with a freshly checked out repository, so cannot comment on if this breaks any tests. From what I can tell this code is untested so it shouldn't break any. 

I tested the changes work as expected locally using existing code.

These functions are not referenced in the main transformers documentation.

@kamilakesbi

### CODE REVIEW

File: src/transformers/pipelines/audio_utils.py [hunk 1]
```diff
@@ -105,6 +100,7 @@ def ffmpeg_microphone_live(
     stream_chunk_s: Optional[int] = None,
     stride_length_s: Optional[Union[Tuple[float, float], float]] = None,
     format_for_conversion: str = "f32le",
+    ffmpeg_input_device: Optional[str] = None,
```
Comment (kamilakesbi) [inline src/transformers/pipelines/audio_utils.py:103]: Could you update the documentation of the `ffmpeg_microphone_live` method?

File: src/transformers/pipelines/audio_utils.py [hunk 2]
```diff
@@ -50,6 +50,7 @@ def ffmpeg_microphone(
     sampling_rate: int,
     chunk_length_s: float,
     format_for_conversion: str = "f32le",
+    ffmpeg_input_device: Optional[str] = None,
```
Comment (kamilakesbi) [inline src/transformers/pipelines/audio_utils.py:53]: It would be nice if we could also add a documentation to the `ffmpeg_microphone` method.

**Review by kamilakesbi — APPROVED** [review]
Hi @jrhe, thanks for working on this! 

The proposed changes look good to me :) 

Could you please update the docs for both `ffmpeg_microphone` and `ffmpeg_microphone_live` methods ?

File: src/transformers/pipelines/audio_utils.py [hunk 3]
```diff
@@ -246,3 +244,13 @@ def _get_microphone_name():
...
+
+
+def _ffmpeg_format_for_platform(platform: str):
+    platform_to_format_map = {"Linux": "alsa", "Darwin": "avfoundation", "Windows": "dshow"}
+    return platform_to_format_map[platform]
+
+
+def _ffmpeg_input_device_for_platform(platform: str):
+    platform_to_input_device_map = {"Linux": "default", "Darwin": ":default", "Windows": _get_microphone_name()}
+    return platform_to_input_device_map[platform]
```
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:256]: The previous if/else logic was better: we only want to run `_get_microphone_name()` when we're running on windows

File: src/transformers/pipelines/audio_utils.py (hunk 2 above)
Comment (amyeroberts) [reply src/transformers/pipelines/audio_utils.py:53]: +1

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! 

We should switch back to guarding `_get_microphone_name` by the platform, +1 to @kamilakesbi's comments on documentation

File: src/transformers/pipelines/audio_utils.py (hunk 3 above)
Comment (jrhe) [reply src/transformers/pipelines/audio_utils.py:256]: Oops! Bit overzealous there, will revert.

**Comment by kamilakesbi** [comment]
@jrhe Have you had time to iterate on this? :)

**Comment by kamilakesbi** [comment]
@jrhe great! thanks for iterating! 

@amyeroberts this should be ready for final review :)

File: src/transformers/pipelines/audio_utils.py [hunk 4]
```diff
@@ -50,9 +50,29 @@ def ffmpeg_microphone(
...
+            avoid resampling later.
+        chunk_length_s (`float` or `int`):
+            The length of the maximum chunk of audio to be sent returned.
+        format_for_conversion (`str`, defaults to `f32le`)
+            The name of the format of the audio samples to be returned by ffmpeg. The standard is `f32le`, `s16le`
+            could also be used.
+        ffmpeg_input_device (`str`, *optional*, defaults to `None`)
+            The indentifier of the input device to be used by ffmpeg (i.e. ffmpeg's '-i' argument). Defaults to `None`,
+            where the default input device will be used. See `https://www.ffmpeg.org/ffmpeg-devices.html#Input-Devices`
+            for how to specify and list input devices.
```
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:72]: ultranit - the pattern in the library docs is not specify if something defaults to `None` and in the description talk about whether the param is 'set' or 'unset'

```suggestion
        ffmpeg_input_device (`str`, *optional*)
            The indentifier of the input device to be used by ffmpeg (i.e. ffmpeg's '-i' argument). If unset,
            the default input device will be used. See `https://www.ffmpeg.org/ffmpeg-devices.html#Input-Devices`
            for how to specify and list input devices.
```

File: src/transformers/pipelines/audio_utils.py [hunk 5]
```diff
@@ -50,9 +50,29 @@ def ffmpeg_microphone(
...
+    Return:
```
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:73]:
```suggestion
    Returns:
```

File: src/transformers/pipelines/audio_utils.py [hunk 6]
```diff
@@ -123,9 +146,13 @@ def ffmpeg_microphone_live(
             The length of the striding to be used. Stride is used to provide context to a model on the (left, right) of
             an audio sample but without using that part to actually make the prediction. Setting this does not change
             the length of the chunk.
-        format_for_conversion (`str`, defalts to `f32le`)
+        format_for_conversion (`str`, defaults to `f32le`)
```
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:149]: nit - in terms of the docstrings, we use "optional" to denote if it's a required argument, or the user can optionally specify it 

```suggestion
        format_for_conversion (`str`, *optional*, defaults to `f32le`)
```

File: src/transformers/pipelines/audio_utils.py [hunk 7]
```diff
@@ -123,9 +146,13 @@ def ffmpeg_microphone_live(
...
             The name of the format of the audio samples to be returned by ffmpeg. The standard is `f32le`, `s16le`
             could also be used.
+        ffmpeg_input_device (`str`, *optional*, defaults to `None`)
+            The indentifier of the input device to be used by ffmpeg (i.e. ffmpeg's '-i' argument). Defaults to `None`,
+            where the default input device will be used. See `https://www.ffmpeg.org/ffmpeg-devices.html#Input-Devices`
+            for how to specify and list input devices.
```
Comment (amyeroberts) [inline src/transformers/pipelines/audio_utils.py:155]:
```suggestion
        ffmpeg_input_device (`str`, *optional*):
            The identifier of the input device to be used by ffmpeg (i.e. ffmpeg's '-i' argument). If unset,
            the default input device will be used. See `https://www.ffmpeg.org/ffmpeg-devices.html#Input-Devices`
            for how to specify and list input devices.
```

**Review by amyeroberts — COMMENTED** [review]
Thanks - looks great! 

Just some small nits on the docstrings

File: src/transformers/pipelines/audio_utils.py (hunk 7 above)
Comment (jrhe) [reply src/transformers/pipelines/audio_utils.py]: Merged manually

**Comment by jrhe** [comment]
Sorry, missed the documentation style guide! Have merged those bits fixed some other bits of documentation (missing colons, and a typo).

**Review by amyeroberts — APPROVED** [review]
Looks great - thanks for adding this!

**Pull request merged**